src/main/java/com/algorand/algosdk/logic/SourceMap.java

```diff
@@ -0,0 +1,112 @@
+package com.algorand.algosdk.logic;
+
+import java.lang.Integer;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+/**
+ * SourceMap class provides parser for source map from  
+ * algod compile endpoint 
+ */
+public class SourceMap {
+
+    public int version;
+    public String file;
+    public String[] sources;
+    public String[] names;
+    public String mappings;
+
+    public HashMap<Integer, Integer> pcToLine;
+    public HashMap<Integer, ArrayList<Integer>> lineToPc;
+
+    public SourceMap(HashMap<String,Object> sourceMap) {
+        int version = (int) sourceMap.get("version");
+        if(version != 3){
+            throw new IllegalArgumentException("Only source map version 3 is supported");
+        }
+        this.version = version;
+
+        this.file = (String) sourceMap.get("file");
+        this.mappings = (String) sourceMap.get("mappings");
+
+        this.lineToPc = new HashMap<>();
+        this.pcToLine = new HashMap<>();
+
+        Integer lastLine = 0;
+        String[] vlqs = this.mappings.split(";");
+        for(int i=0; i<vlqs.length; i++){
+            ArrayList<Integer> vals = VLQDecoder.decodeSourceMapLine(vlqs[i]);
+
+        	// If the vals length >= 3 the lineDelta
+            if(vals.size() >= 3){
+                lastLine = lastLine + vals.get(2);
+            }
+
+            if(!this.lineToPc.containsKey(lastLine)){
+                this.lineToPc.put(lastLine, new ArrayList<Integer>());
+            }
+
+            ArrayList<Integer> currList = this.lineToPc.get(lastLine);
+            currList.add(i);
+            this.lineToPc.put(lastLine, currList);
+
+            this.pcToLine.put(i, lastLine);
+        }
+
+    }
+
+    public Integer getLineForPc(Integer pc) {
+        if(!this.pcToLine.containsKey(pc)){
+            return 0;
+        }
+        return this.pcToLine.get(pc);
+    }
+
+    public ArrayList<Integer> getPcsForLine(Integer line) {
+        if(!this.pcToLine.containsKey(line)){
+            return new ArrayList<Integer>();
+        }
+        return this.lineToPc.get(line);
+    }
+
+    private static class VLQDecoder {
```

Comment:
@barnjamin Checking:  Is there a citation that be added inline to help other maintainers understand where the VLQ decoding implementation came from?

I'm assuming (perhaps incorrectly) that it's based on one of the other SDKs.

Reply:
yes i took it from the go sdk, i can add some documentation to provide context

Reply:
@barnjamin Thanks for the note - looks reasonable, resolving.